src/unix/lwt_domain.ml

```diff
@@ -0,0 +1,67 @@
+open Lwt.Infix
+
+module C = Domainslib.Chan
+module T = Domainslib.Task
+
+type pool = Domainslib.Task.pool
+
+let setup_pool ?name num_additional_domains =
+    T.setup_pool ?name ~num_additional_domains:num_additional_domains ()
+
+
+let teardown_pool = T.teardown_pool
+
+let lookup_pool = T.lookup_pool
+
+let get_num_domains = T.get_num_domains
+
+let init_result = Result.Error (Failure "Lwt_domain.detach")
+
+let detach pool f args =
+  if (get_num_domains pool = 1) then
+    Lwt.wrap1 f args
+  else begin
+    let result = ref init_result in
+    let task () =
+      result := try Result.Ok (f args) with exn -> Result.Error exn
+    in
+    let waiter, wakener = Lwt.wait () in
+    let id =
+      Lwt_unix.make_notification ~once:true
+        (fun () -> Lwt.wakeup_result wakener !result)
+    in
+    let _ = T.async pool (fun _ -> task ();
+    Lwt_unix.send_notification id) in
+    waiter
+  end
+
+(* +-----------------------------------------------------------------+
+   | Running Lwt threads in the main domain                          |
+   +-----------------------------------------------------------------+ *)
+
+(* Jobs to be run in the main domain *)
+let jobs = C.make_unbounded ()
+let job_done = C.make_bounded 0
+let job_notification =
+  Lwt_unix.make_notification
+    (fun () ->
+      let thunk = C.recv jobs in
+      ignore (thunk ()))
+
+let run_in_main f =
+  let res = ref init_result in
+  let job () =
+    Lwt.try_bind f
+      (fun ret -> Lwt.return (Result.Ok ret))
+      (fun exn -> Lwt.return (Result.Error exn)) >>= fun result ->
+    res := result;
+    C.send job_done 1;
+    Lwt.return_unit
+  in
```

Comment:
```suggestion
  let job () =
    let open Lwt.Syntax in
    let+ result = Lwt.try_bind f Lwt_result.return Lwt_result.fail in
    res := result;
    C.send job_done ()
  in
```

Taking advantage of some slightly newer features in Lwt, like the [`Lwt_result`](http://ocsigen.org/lwt/latest/api/Lwt_result) module and letops. Also I think this is the only use of `Lwt.Infix` in this file so that open is not needed any more.

Also I am suggesting to change the `job_done` channel to type `unit` as its payload carries no meaning.

Reply:
It's somewhat annoying that `Lwt_result.catch` lets some exceptions escape by taking a promise rather than a thunk as argument, otherwise it'd have been perfect.

We could use `let+ result = Lwt_result.catch (Lwt.apply f ()) in ..` but that's not more readable.

Reply:
>It's somewhat annoying that Lwt_result.catch lets some exceptions escape by taking a promise rather than a thunk as argument

Are there any examples of some exceptions escaping `Lwt_result.catch`? The rather simple ones I tried out seemed to work as expected. Also, is this something an user needs to be aware of?

Reply:
The issue only appears for somewhat corner cases:

```
# let f x = if x < 0 then raise (Invalid_argument "") else Lwt.pause () >>= fun () -> Lwt.return x ;;
val f : int -> int Lwt.t = <fun>
# Lwt_result.catch (f (-1)) ;;
Exception: Invalid_argument "".
```

The issue is that expressions that raise an exception can be given the type of promises. Compare to `catch` which takes a `unit -> 'a t` instead of an `'a t` which allows it to install a `try`-`catch` around it.

But fixing that is out of the scope of this MR.

-----------------

I find the current form of the code satisfying. I'm resolving because this is a somewhat cosmetic suggestion. @yawaramin feel free to open an MR against this branch (or master if this branch is merged) to get this and other improvements merged.